pybop/observers/unscented_kalman.py

```diff
@@ -0,0 +1,309 @@
+
+    Parameters
+    ----------
+    x0 : np.ndarray
+        The initial state vector
+    P0 : np.ndarray
+        The initial covariance matrix
+    Rp : np.ndarray
+        The covariance matrix of the process noise
+    Rm : np.ndarray
+        The covariance matrix of the measurement noise
+    f : callable
+        The state transition function
+    h : callable
+        The measurement function
+    """
+
+    def __init__(
+        self,
+        x0: np.ndarray,
+        P0: np.ndarray,
+        Rp: np.ndarray,
+        Rm: np.ndarray,
+        f: callable,
+        h: callable,
+    ) -> None:
+        self.x = x0
+        self.S = linalg.cholesky(P0)
+        self.sqrtRp = linalg.cholesky(Rp)
+        self.sqrtRm = linalg.cholesky(Rm)
+        self.alpha = 1e-3
+        self.beta = 2
+        self.f = f
+        self.h = h
+
+    @staticmethod
+    def gen_sigma_points(
+        x: np.ndarray, S: np.ndarray, alpha: float, beta: float
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Generates the sigma points for the unscented transform
+
+        Parameters
+        ----------
+        x : np.ndarray
+            The state vector
+        S : np.ndarray
+            The square root of the covariance matrix
+        alpha : float
+            The spread of the sigma points. Typically 1e-4 < alpha < 1
+        beta : float
+            The prior knowledge of the distribution. Typically 2 for a Gaussian distribution
+
+        Returns
+        -------
+        List[np.ndarray]
+            The sigma points
+        List[float]
+            The weights of the sigma points
+        List[float]
+            The weights of the covariance points
+        """
+        kappa = 1.0
+        L = len(x)
+        sigma = alpha**2 * (L + kappa) - L
+        eta = np.sqrt(L + sigma)
+        wm_0 = sigma / (L + sigma)
+        wc_0 = wm_0 + (1 - alpha**2 + beta)
+        points = np.hstack(
+            [x]
+            + [x + eta * S[:, i].reshape(-1, 1) for i in range(L)]
+            + [x - eta * S[:, i].reshape(-1, 1) for i in range(L)]
+        )
+        w_m = np.array([wm_0] + [(1 - wm_0) / (2 * L)] * (2 * L))
+        w_c = np.array([wc_0] + [(1 - wc_0) / (2 * L)] * (2 * L))
+        return (points, w_m, w_c)
+
+    @staticmethod
+    def unscented_transform(
+        sigma_points: np.ndarray, w_m: np.ndarray, w_c: np.ndarray, sqrtR: np.ndarray
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Performs the unscented transform
+
+        Parameters
+        ----------
+        sigma_points : List[SigmaPoint]
+            The sigma points
+        sqrtR : np.ndarray
+            The square root of the covariance matrix
+
+        Returns
+        -------
+        Tuple[np.ndarray, np.ndarray]
+            The mean and covariance of the sigma points
+        """
+        x = np.sum(w_m * sigma_points, axis=1).reshape(-1, 1)
```

Comment:
Can this be replaced with a dot product?

```
x = np.dot(w_m, sigma_points).reshape(-1,1)
```

Reply:
perhaps? you are relying on `w_m` being broadcasted correctly which the `*` does atm. I can give `dot` a go and see if it works

Reply:
nope, `dot` does not do the broadcasting :( 

```
ValueError: shapes (2,) and (4,2) not aligned: 2 (dim 0) != 4 (dim 0)
```

Reply:
Makes sense, thought it was worth a double-check. Thanks for investigating